File: openspec/changes/build-bloomfilters-from-scratch-and-learn/BloomFilter.py
class BloomFilter:
    def __init__(self, size, hash_functions):
        """
        Initialize the Bloom filter.

        :param size: Number of bits in the Bloom filter.
        :param hash_functions: List of hash functions to use.
        """
        self.size = size
        self.bit_array = [0] * size
        self.hash_functions = hash_functions

    def add(self, item):
        """
        Add an item to the Bloom filter.

        :param item: The item to add.
        """
        for hash_function in self.hash_functions:
            index = hash_function(item) % self.size
            self.bit_array[index] = 1

    def contains(self, item):
        """
        Check if an item is in the Bloom filter.

        :param item: The item to check.
        :return: True if the item is probably in the filter, False if it is definitely not.
        """
        for hash_function in self.hash_functions:
            index = hash_function(item) % self.size
            if self.bit_array[index] == 0:
                return False
        return True

# Example usage
if __name__ == "__main__":
    import hashlib

    def hash_fn1(x):
        return int(hashlib.md5(x.encode()).hexdigest(), 16)

    def hash_fn2(x):
        return int(hashlib.sha1(x.encode()).hexdigest(), 16)

    def hash_fn3(x):
        return int(hashlib.sha256(x.encode()).hexdigest(), 16)

    # Create a Bloom filter with 1000 bits and 3 hash functions
    bloom_filter = BloomFilter(1000, [hash_fn1, hash_fn2, hash_fn3])

    # Add items to the Bloom filter
    bloom_filter.add("apple")
    bloom_filter.add("banana")

    # Check membership
    print(bloom_filter.contains("apple"))  # True
    print(bloom_filter.contains("banana"))  # True
    print(bloom_filter.contains("cherry"))  # False

File: openspec/changes/build-bloomfilters-from-scratch-and-learn/BloomFilter.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.BitSet;
import java.util.function.Function;

public class BloomFilter {
    private final BitSet bitSet;
    private final int size;
    private final Function<String, Integer>[] hashFunctions;

    @SafeVarargs
    public BloomFilter(int size, Function<String, Integer>... hashFunctions) {
        this.size = size;
        this.bitSet = new BitSet(size);
        this.hashFunctions = hashFunctions;
    }

    public void add(String item) {
        for (Function<String, Integer> hashFunction : hashFunctions) {
            int index = Math.abs(hashFunction.apply(item) % size);
            bitSet.set(index);
        }
    }

    public boolean contains(String item) {
        for (Function<String, Integer> hashFunction : hashFunctions) {
            int index = Math.abs(hashFunction.apply(item) % size);
            if (!bitSet.get(index)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) throws NoSuchAlgorithmException {
        // Define hash functions
        Function<String, Integer> hashFn1 = createHashFunction("MD5");
        Function<String, Integer> hashFn2 = createHashFunction("SHA-1");
        Function<String, Integer> hashFn3 = createHashFunction("SHA-256");

        // Create a Bloom filter with 1000 bits and 3 hash functions
        BloomFilter bloomFilter = new BloomFilter(1000, hashFn1, hashFn2, hashFn3);

        // Add items to the Bloom filter
        bloomFilter.add("apple");
        bloomFilter.add("banana");

        // Check membership
        System.out.println(bloomFilter.contains("apple"));  // True
        System.out.println(bloomFilter.contains("banana")); // True
        System.out.println(bloomFilter.contains("cherry")); // False
    }

    private static Function<String, Integer> createHashFunction(String algorithm) throws NoSuchAlgorithmException {
        MessageDigest messageDigest = MessageDigest.getInstance(algorithm);
        return (String input) -> {
            messageDigest.update(input.getBytes(StandardCharsets.UTF_8));
            byte[] digest = messageDigest.digest();
            int hash = 0;
            for (byte b : digest) {
                hash = (hash << 8) | (b & 0xFF);
            }
            return hash;
        };
    }
}